pkg/controller/master/virtualmachine/vmi_network_controller.go

```diff
@@ -71,14 +70,14 @@ func (h *VMNetworkController) updateVMDefaultNetworkMacAddress(vmi *kubevirtv1.V
 		}
 	}
 
-	if reflect.DeepEqual(vmCopy.Spec.Template.Spec.Domain.Devices, vm.Spec.Template.Spec.Domain.Devices) {
-		return nil
-	}
-
-	logrus.Debugf("update default network mac address of the vm: %s", vmi.Name)
+	// note: this function is related to vm/vmi, should always run on vmi change
 	if err := h.regenerateControllerRevision(vmi, vm); err != nil {
 		return err
 	}
+
+	if equality.Semantic.DeepEqual(vmCopy.Spec.Template.Spec.Domain.Devices, vm.Spec.Template.Spec.Domain.Devices) {
```

Comment:
Just curious, why do we use reflect.DeepEqual here ?

Reply:
there are pointers in the struct, `equality.Semantic.DeepEqual` checks the value of the pointer

Reply:
Correct me if I'm wrong, but I tried reflect.DeepEqual. Looks like it also compares the value of the pointer.
https://go.dev/play/p/56N4QZIKOKZ?v=goprev

I'm not trying to debate something, just want to clarify the timing of using `equality.Semantic.DeepEqual` / `reflect.DeepEqual`. As in the harvester repo we mostly use `reflect.DeepEqual` for comparison.

Reply:
Just a quick survey, what I know is that there are differences in comparison for certain data structures. In Kubernetes, equality.Semantic.DeepEqual returns true when comparing nil and empty maps, meaning they are considered equal. https://pkg.go.dev/k8s.io/apimachinery/pkg/api/equality

However, this is not the case with reflect.DeepEqual, where nil and empty maps are not considered equal.
see https://go.dev/play/p/U8zt1G34ztR?v=goprev

I'm not sure if this is something we need to aware of

Reply:
Good question and investigation. Thanks.

Honestly, I did not dive so much on this topic. The reason to use `equality.Semantic.DeepEqual` instead of `reflect` is: the `equality` package has been included in this file but `reflect` was added later. And it was used in another function to check as such `!equality.Semantic.DeepEqual(revisionSpec.Spec.Template, vm.Spec.Template)`

Reply:
Further tested with following code, both `reflect` and `equality` work well on this conext. Preferring `equality` is mainly because it has been used in this file.

My above answer `there are pointers in the struct, equality.Semantic.DeepEqual checks the value of the pointer` was not fully accurate, sorry for the inconvenience.

![image](https://github.com/user-attachments/assets/f8e3f18a-2455-4115-a12d-eefea99e1d4d)

Thanks again.